### PR TITLE
Use zip deploy to upload jar related files

### PR DESCRIPTION
**Pull request by SummerSun**
#1969

### CODE REVIEW

File: Utils/azuretools-core/src/com/microsoft/azuretools/azurecommons/util/FileUtil.java [hunk 1]
```diff
@@ -105,4 +107,37 @@ public static boolean isNullOrEmpty(final String name) {
...
 	}
 
+    /**
+     * Utility method to zip a file.
+     * @param inputStream
+     * @param fileName
+     * @return zip file
+     * @throws IOException
+     */
+    public static File zipFile(InputStream inputStream, String fileName) throws IOException {
```
Comment (jdneo) [inline Utils/azuretools-core/src/com/microsoft/azuretools/azurecommons/util/FileUtil.java]: Can we use third-party lib to handle this? For example: https://github.com/zeroturnaround/zt-zip

Comment (SummerSun) [reply Utils/azuretools-core/src/com/microsoft/azuretools/azurecommons/util/FileUtil.java]: Adding external dependency is not suggested and it needs a complicated process. All zip related operations are provided by util functions in this project.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/WebAppRunState.java [hunk 2]
```diff
@@ -184,41 +183,71 @@ private WebApp getWebAppAccordingToConfiguration(@NotNull RunProcessHandler proc
...
-        if (webAppSettingModel.isCreatingNew() && Comparing.equal(fileType, MavenConstants.TYPE_JAR)) {
-            processHandler.setText(UPLOADING_WEB_CONFIG);
-            try (InputStream webConfigInput = getClass()
-                    .getResourceAsStream(WEB_CONFIG_PACKAGE_PATH)) {
-                return uploadFileToFtp(ftp, WEB_CONFIG_FTP_PATH, webConfigInput, processHandler);
-            }
+    private void prepareWebConfig(@NotNull final WebApp webApp, @NotNull final RunProcessHandler processHandler,
+                                  @NotNull final Map<String, String> telemetryMap) throws IOException {
+        processHandler.setText(UPLOADING_WEB_CONFIG);
+        try (InputStream webConfigInputStream = getClass().getResourceAsStream(WEB_CONFIG_PACKAGE_PATH)) {
```
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/WebAppRunState.java]: Why do we pass InputStream down ?

File: Utils/azuretools-core/src/com/microsoft/azuretools/azurecommons/util/FileUtil.java (hunk 1 above)
Comment (Eskibear) [reply Utils/azuretools-core/src/com/microsoft/azuretools/azurecommons/util/FileUtil.java]: > ZipUtil.pack(new File("/tmp/demo"), new File("/tmp/demo.zip"));

I think the signature `(File src, File dst)` is more clear if I want to compress some files. Is it possible to do in that way?

File: Utils/azuretools-core/src/com/microsoft/azuretools/azurecommons/util/FileUtil.java [hunk 3]
```diff
@@ -105,4 +107,37 @@ public static boolean isNullOrEmpty(final String name) {
 		return isValid;
 	}
 
+    /**
+     * Utility method to zip a file.
+     * @param inputStream
```
Comment (Eskibear) [inline Utils/azuretools-core/src/com/microsoft/azuretools/azurecommons/util/FileUtil.java]: From the JavaDoc, I don't understand its usage.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/WebAppRunState.java [hunk 4]
```diff
@@ -184,41 +183,71 @@ private WebApp getWebAppAccordingToConfiguration(@NotNull RunProcessHandler proc
...
+        int uploadCount = 0;
+        while (uploadCount < UPLOADING_MAX_TRY) {
+            uploadCount += 1;
+            try {
+                webapp.zipDeploy(FileUtil.zipFile(inputStream, fileName));
+                processHandler.setText(UPLOADING_SUCCESSFUL);
+                return uploadCount;
+            } catch (Exception e) {
+                processHandler.setText(
+                    String.format("Upload file via zip deploy met exception: %s, retry immediately...", e.getMessage()));
```
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/WebAppRunState.java]: Take care of the error message.

Comment (Eskibear) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/WebAppRunState.java]: And better to convert the message template as a field constant.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/WebAppRunState.java [hunk 5]
```diff
@@ -184,41 +183,71 @@ private WebApp getWebAppAccordingToConfiguration(@NotNull RunProcessHandler proc
...
+            }
         }
+        throw new IOException(String.format("Upload failed after %d times of retry.", UPLOADING_MAX_TRY));
```
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/WebAppRunState.java]: lso take care of the error message here.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/WebAppRunState.java (hunk 4 above)
Comment (SummerSun) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/WebAppRunState.java]: Any suggestion for the error message? For error message that is not reused, I prefer to use a pure string where it is used for better readability.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/WebAppRunState.java (hunk 5 above)
Comment (SummerSun) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/WebAppRunState.java]: Ang suggestion for the error message?

**Comment by SummerSun** [comment]
Since the `zipDeploy` will delete files that are no longer in the build ([details](https://github.com/projectkudu/kudu/wiki/Deploying-from-a-zip-file)) if we keep the design that deploys `web.config` only when creating a new web app, the file will be deleted any time we deploy a jar file. 

Will update it to always deploy `jar` file and `web.config` in the same time in next PR. Issue #1979 is created to track the issue.

File: Utils/azuretools-core/src/com/microsoft/azuretools/azurecommons/util/FileUtil.java [hunk 6]
```diff
@@ -108,18 +109,26 @@ public static boolean isNullOrEmpty(final String name) {
...
+     * Utility method to zip the given source file to the destination file.
+     * @param sourceFile source file
+     * @param targetZipFile ZIP file that will be created or overwritten
      */
-    public static File zipFile(InputStream inputStream, String fileName) throws IOException {
-        final File tempZip = File.createTempFile(String.valueOf(new Date().getTime()), ".zip");
-        final FileOutputStream fos = new FileOutputStream(tempZip);
+    public static void zipFile(final @NotNull File sourceFile, final @NotNull File targetZipFile) throws Exception {
+        if (!sourceFile.exists()) {
+            throw new Exception("The source file to zip does not exist.");
```
Comment (Eskibear) [inline Utils/azuretools-core/src/com/microsoft/azuretools/azurecommons/util/FileUtil.java]: https://docs.oracle.com/javase/7/docs/api/java/io/FileNotFoundException.html

Comment (Eskibear) [reply Utils/azuretools-core/src/com/microsoft/azuretools/azurecommons/util/FileUtil.java]: Anyway I think we can do it later, just don't forget it.

Comment (SummerSun) [reply Utils/azuretools-core/src/com/microsoft/azuretools/azurecommons/util/FileUtil.java]: Track it in this issue https://github.com/Microsoft/azure-tools-for-java/issues/1983

**Pull request merged**